connectivity/service/src/main/java/org/eclipse/ditto/connectivity/service/messaging/mqtt/hivemq/AbstractHiveMqttClientFactory.java

```diff
@@ -56,25 +57,38 @@ protected AbstractHiveMqttClientFactory(final Supplier<SshTunnelState> tunnelSta
     // duplicate code unavoidable because there is no common interface
     // between Mqtt3SimpleAuthBuilder.Nested and Mqtt5SimpleAuthBuilder.Nested
     @SuppressWarnings("Duplicates")
-    void configureSimpleAuth(final Mqtt3SimpleAuthBuilder.Nested<?> simpleAuth, final Connection connection) {
-        final Optional<String> possibleUsername = connection.getUsername();
-        final Optional<String> possiblePassword = connection.getPassword();
-        if (possibleUsername.isPresent() && possiblePassword.isPresent()) {
-            simpleAuth.username(possibleUsername.get())
-                    .password(possiblePassword.get().getBytes(StandardCharsets.UTF_8))
+    void configureSimpleAuth(final Mqtt3SimpleAuthBuilder.Nested<?> simpleAuth, final Connection connection, final boolean doubleDecodingEnabled) {
+        final Optional<String> username =
+                connection.getUsername().map(u -> doubleDecodingEnabled ? tryDecodeUriComponent(u) : u);
+        final Optional<String> password =
+                connection.getPassword().map(p -> doubleDecodingEnabled ? tryDecodeUriComponent(p) : p);
+        if (username.isPresent() && password.isPresent()) {
+            simpleAuth.username(username.get())
+                    .password(password.get().getBytes(StandardCharsets.UTF_8))
                     .applySimpleAuth();
         }
     }
 
+    private static String tryDecodeUriComponent(final String string) {
```

Comment:
why was the decoding logic moved out of the model? Wouldn't it be possible to keep it in there and instead pass the boolean to the "getPassword" method? Now we have this logic duplicated at three places.

Reply:
You're right!
https://github.com/eclipse/ditto/pull/1346/commits/0417fd4b5e6a18d04fb81f7cedd0ad8a2a9f152b